### PR TITLE
Add build/compile actions and improved optional action support

### PR DESCRIPTION
**Pull request by SanjulaGanepola**
### Changes
- Added inline actions on the project to run a build or action
- Add right-click actions on the project to set the build/compile or launch action setup
- Added action on the editor context menu and file/directory context menus in the file explorer to run a compile
- Added the build and compile actions to the command palette
- Added build (`ctrl+shift+b` or `cmd+shift+b`) and compile (`ctrl+shift+c` or `cmd+shift+c`) keybinds
- Added build and compile events

image

image

image

### CODE REVIEW

**Comment by SanjulaGanepola** [comment]
This PR requires PR #97 to be merged in first. It also depends on https://github.com/halcyon-tech/vscode-ibmi/pull/1484

File: src/iproject.ts [hunk 1]
```diff
@@ -427,6 +427,88 @@ export class IProject {
...
+  /**
+   * Run the project's build or compile command.
+   * 
+   * @param isBuild True for build command and false for compile command.
+   */
+  public async runBuildOrCompileCommand(isBuild: boolean) {
+    const unresolvedState = await this.getUnresolvedState();
+
+    if (unresolvedState) {
+      const command = isBuild ? unresolvedState.buildCommand : unresolvedState.compileCommand;
```
Comment (edmundreinhardt) [inline src/iproject.ts:439]: shouldn't we resolve variables in this command string?

File: src/views/projectExplorer/index.ts [hunk 2]
```diff
@@ -71,6 +71,92 @@ export default class ProjectExplorer implements TreeDataProvider<ProjectExplorer
...
+      commands.registerCommand(`vscode-ibmi-projectexplorer.projectExplorer.setBuildCommand`, async (element: Project) => {
+        if (element) {
+          const iProject = ProjectManager.get(element.workspaceFolder);
+
+          if (iProject) {
+            const unresolvedState = await iProject.getUnresolvedState();
+
+            if (unresolvedState) {
+              const command = await window.showInputBox({
+                prompt: l10n.t('Enter build command'),
```
Comment (edmundreinhardt) [inline src/views/projectExplorer/index.ts:97]: Should we show what variables can be set.  Merlin allowed some useful one like active file etc.

File: src/views/projectExplorer/index.ts [hunk 3]
```diff
@@ -71,6 +71,92 @@ export default class ProjectExplorer implements TreeDataProvider<ProjectExplorer
...
+      commands.registerCommand(`vscode-ibmi-projectexplorer.projectExplorer.setCompileCommand`, async (element: Project) => {
+        if (element) {
+          const iProject = ProjectManager.get(element.workspaceFolder);
+
+          if (iProject) {
+            const unresolvedState = await iProject.getUnresolvedState();
+
+            if (unresolvedState) {
+              const command = await window.showInputBox({
+                prompt: l10n.t('Enter compile command'),
```
Comment (edmundreinhardt) [inline src/views/projectExplorer/index.ts:138]: definitely need  variable for file name documented here

**Review by edmundreinhardt — CHANGES_REQUESTED** [review]

File: src/iproject.ts (hunk 1 above)
Comment (SanjulaGanepola) [reply src/iproject.ts:439]: The `code-for-ibmi.runAction` command provided by Code4i will automatically handle the resolving of variables.

File: src/views/projectExplorer/index.ts (hunk 3 above)
Comment (SanjulaGanepola) [reply src/views/projectExplorer/index.ts:138]: Same as above

File: src/views/projectExplorer/index.ts [hunk 4]
```diff
@@ -94,7 +94,12 @@ export default class ProjectExplorer implements TreeDataProvider<ProjectExplorer
 
             if (unresolvedState) {
               const command = await window.showInputBox({
-                prompt: l10n.t('Enter build command'),
+                prompt: l10n.t('Enter build command (')
```
Comment (edmundreinhardt) [inline src/views/projectExplorer/index.ts:97]: please have mercy on translator and make this one string

File: src/views/projectExplorer/index.ts [hunk 5]
```diff
@@ -135,7 +140,12 @@ export default class ProjectExplorer implements TreeDataProvider<ProjectExplorer
 
             if (unresolvedState) {
               const command = await window.showInputBox({
-                prompt: l10n.t('Enter compile command'),
+                prompt: l10n.t('Enter compile command (')
+                  + l10n.t('{0} resolves to the base file name being edited. ', '{filename}')
```
Comment (edmundreinhardt) [inline src/views/projectExplorer/index.ts:144]: please have mercy on translator and make this one string

**Review by edmundreinhardt — CHANGES_REQUESTED** [review]

**Pull request merged**